cpp/src/slice2cpp/Gen.cpp

```diff
@@ -1826,167 +1822,120 @@ Slice::Gen::ProxyVisitor::visitOperation(const OperationPtr& p)
     H << nl;
     H << deprecateSymbol;
     H << "::std::function<void()>";
-    H << nl << name << "Async(";
-    H.useCurrentPosAsIndent();
-    if(!inParamsDecl.empty())
-    {
-        for(vector<string>::const_iterator q = inParamsDecl.begin(); q != inParamsDecl.end(); ++q)
-        {
-            if(q != inParamsDecl.begin())
-            {
-                H << " ";
-            }
-
-            H << *q << ",";
-        }
-        H << nl;
-    }
 
-    H << "::std::function<void" << spar << lambdaOutParams << epar << "> " << responseParam << ",";
-    H << nl << "::std::function<void(::std::exception_ptr)> " << exParam << " = nullptr,";
-    H << nl << "::std::function<void(bool)> " << sentParam << " = nullptr,";
-    H << nl << contextDecl << ") const;";
+    // TODO: need "nl" version of spar/epar
+    H << nl << name << "Async" << spar;
+    H.useCurrentPosAsIndent();
+    H << inParamsDecl;
 
+    H << lambdaResponse + " " + responseParam;
+    H << "::std::function<void(::std::exception_ptr)> " + exParam + " = nullptr";
+    H << "::std::function<void(bool)> " + sentParam + " = nullptr";
+    H << contextDecl << epar << " const;";
     H.restoreIndent();
 
     C << sp;
     C << nl << "::std::function<void()>";
-    C << nl << scoped << name << "Async(";
+    C << nl << scoped << name << "Async" << spar;
     C.useCurrentPosAsIndent();
-    if (!inParamsImplDecl.empty())
+    C << inParamsImplDecl;
+    C << lambdaResponse + " response";
+    C << "::std::function<void(::std::exception_ptr)> ex";
+    C << "::std::function<void(bool)> sent";
+    C << "const ::Ice::Context& context" << epar << " const";
+    C.restoreIndent();
+
+    C << sb;
+    if (lambdaOutParams.size() > 1)
     {
-        for (vector<string>::const_iterator q = inParamsImplDecl.begin(); q != inParamsImplDecl.end(); ++q)
-        {
-            if (q != inParamsImplDecl.begin())
-            {
-                C << " ";
-            }
-            C << *q << ",";
-        }
-        C << nl;
+        C << nl << "auto responseCb = [response = ::std::move(response)]("
+            << lambdaT << "&& result)";
+        C << sb;
+        C << nl << "::std::apply(::std::move(response), ::std::move(result));";
+        C << eb << ";";
     }
 
-    C << "::std::function<void " << spar << lambdaOutParams << epar << "> response,";
-    C << nl << "::std::function<void(::std::exception_ptr)> ex,";
-    C << nl << "::std::function<void(bool)> sent,";
-    C << nl << "const ::Ice::Context& context) const";
-    C.restoreIndent();
+    C << nl << "return ::IceInternal::makeLambdaOutgoing<" << lambdaT << ">" << spar;
+
+    C << "std::move(" + (lambdaOutParams.size() > 1 ? string("responseCb") : "response") + ")"
```

Comment:
We should use the stream operation here.

Reply:
The rule is simple: within a spar/epar (and similar), using the stream operator adds a separator (comma with spar/epar). So when we don't want a comma, we must use "+".